py-rse/package-py.Rmd

```diff
@@ -482,450 +173,562 @@ without affecting the main installation (Figure \@ref(fig:py-rse-package-py-vir
 
-Ah ha - we're missing a dependency. We use `pytest` for approximate
-value comparison.  How best should we document this requirement?
+```shell
+$ pip install twine
+```
+and now we can upload our distributions from the `dist/` folder:
 
-Since a project may depend on many packages,
-developers frequently put a list of those dependencies in a file called `requirements.txt`.
-`pip install -r requirements.txt` will then install the dependencies listed in that file.
-(The file can be called anything,
-but everyone uses `requirements.txt`,
-so you should too.)
-This file can just list package names,
-or it can specify exact versions, minimum versions, etc.:
+```shell
+$ twine upload --repository-url https://test.pypi.org/legacy/ dist/*
+Enter your username: amirakhan
+Enter your passowrd: *********
+Uploading distributions to https://test.pypi.org/legacy/
+Uploading zipfs_law-0.1.tar.gz
+100%|█████████████████| 5.59k/5.59k [00:01<00:00, 3.27kB/s]
 
-```text
-request
-scipy==1.1.0
-tdda>=1.0
+View at:
+https://test.pypi.org/project/zipfs-law/0.1/
 ```
-If you want to create a file like this,
-`pip freeze` will print the exact versions of all installed packages.
-`pip freeze` provides the same output as `pip list`, but in less human-reader-friendly
-way and a more easily parsed by computer program format.
 
-So let's install the dependency, ensure that it works, and then create the
-`requirements.txt` file:
+The `--repository-url` allows us to specify the test PyPI repository, and we
+have now uploaded both types of distribution, allowing users with different
+environments to use the wheel distribution if their systems allow it, or
+the source distribution otherwise.
+
+We can now test this with a virtual environment and installing with pip:
 
-```text
-(test_zipfpy) $ pip uninstall zipfpy
-(test_zipfpy) $ pip install pytest
-Collecting pytest
-...[stuff omitted]...
-Installing collected packages: py, zipp, importlib-metadata, pluggy, wcwidth, pyparsing, six, packaging, attrs, more-itertools, atomicwrites, pytest
-Successfully installed atomicwrites-1.3.0 attrs-19.1.0 importlib-metadata-0.18 more-itertools-7.0.0 packaging-19.0 pluggy-0.12.0 py-1.8.0 pyparsing-2.4.0 pytest-4.6.3 six-1.12.0 wcwidth-0.1.7 zipp-0.5.1
-(test_zipfpy) $ pip freeze > requirements.txt
-(test_zipfpy) $ cat requirements.txt
-atomicwrites==1.3.0
-attrs==19.1.0
-importlib-metadata==0.18
-more-itertools==7.0.0
-packaging==19.0
-pluggy==0.12.0
-py==1.8.0
-pyparsing==2.4.0
-pytest==4.6.3
-six==1.12.0
-wcwidth==0.1.7
-zipp==0.5.1
+```shell
+$ conda create -n zipfs-law-test
+$ conda activate zipfs-law-test
+$ pip install --index-url https://test.pypi.org/simple zipfs_law
 ```
 
-(Note that the version numbers, current at the time of writing, will change).
+Success!  Once we are happy with how our package appears in testpypi
+(including its [project page](https://test.pypi.org/project/zipfpy/)),
+we can go through the same process to put it on the main pypi repository.
```

Comment:
Should be https://test.pypi.org/project/zipfs-law (although I like the name zipfpy!)